std/random.d

```diff
@@ -1353,25 +1638,119 @@ Twister has `2 ^^ 19937 - 1` distinct states but after `seed(uint)` is
+                        atomicStore(_globalGamma, gamma);
+                    }
+                }
+                assert(gamma);
+            }
+            // Stafford Mix05
+            ulong x = atomicOp!"+="(_globalSeed, gamma);
+            x = (x ^ (x >>> 31)) * 0x79c1_35c1_674b_9addUL;
+            x = (x ^ (x >>> 29)) * 0x54c7_7c86_f691_3e45UL;
+            return x ^ (x >>> 30);
+        }
+        else
         {
-            uint threadID = cast(uint) cast(void*) Thread.getThis();
-            rand.seed((getpid() + threadID) ^ cast(uint) MonoTime.currTime.ticks);
-            seeded = true;
+            shared static uint _globalSeedLow;
+            shared static uint _globalSeedHigh;
+            shared static uint _globalGammaLow;
+            shared static uint _globalGammaHigh;
+            uint gammaLow = atomicLoad!(MemoryOrder.raw)(_globalGammaLow);
+            if (!gammaLow)
+            {
+                synchronized
+                {
+                    gammaLow = atomicLoad!(_globalGammaLow);
+                    if (!gammaLow)
+                    {
+                        ulong[2] s = oneTimeInitSeedAndGamma();
+                        atomicStore(_globalSeedHigh, cast(uint) (s[0] >>> 32));
+                        atomicStore(_globalSeedLow, cast(uint) s[0]);
+                        gammaLow = cast(uint) s[1];
+                        uint h = cast(uint) (s[1] >>> 32);
+                        if (h == 0) // Should be a constraint violation, but check anyway.
+                            h = 1;
+                        atomicStore(_globalGammaHigh, h);
+                        atomicStore(_globalGammaLow, gammaLow);
+                    }
+                }
+                assert(gammaLow);
+            }
+            ulong gammaHigh = atomicLoad!(MemoryOrder.raw)(_globalGammaHigh);
+            if (!gammaHigh)
+            {
+                 gammaHigh = atomicLoad(_globalGammaHigh);
+                 assert(gammaHigh);
+            }
+            // Under contention one thread can update the low half of
+            // the global seed, then several other threads could update
+            // the low half and the high half of the global seed, then
+            // the first thread could update the high half of the global
+            // seed. This is completely fine: the same value will not be
+            // seen by more than one thread and updates will not be lost.
+            uint low = atomicOp!"+="(_globalSeedLow, gammaLow);
+            uint high = atomicOp!"+="(_globalSeedHigh, gammaHigh + uint(low < gammaLow));
+            // Stafford Mix05
+            ulong x = (cast(ulong) high) << 32 + cast(ulong) low;
+            x = (x ^ (x >>> 31)) * 0x79c1_35c1_674b_9addUL;
+            x = (x ^ (x >>> 29)) * 0x54c7_7c86_f691_3e45UL;
+            return x ^ (x >>> 30);
         }
-        rand.popFront();
-        return cast(uint) (MonoTime.currTime.ticks ^ rand.front);
     }
 }
 
```

Comment:
How about adding a lit test that a `unpredictableSeedOf` (or an array of it), is non-zero?

Reply:
added